File: Programmers/Python/Programming_Maester/게임맵최단거리/게임맵최단거리.py
from collections import deque

def bfs(start,maps):

    direct = [[-1,0],[0,-1],[1,0],[0,1]]
    queue = deque()
    queue.append(start)

    while queue:
        y, x , cnt= queue.popleft()
        maps[y][x] = 0

        for dy, dx in direct:
        
            next_x , next_y =  x + dx, y + dy
            
            if next_y == len(maps) -1  and next_x == len(maps[0]) -1 :
                return cnt + 1
            
            if 0 <= next_y < len(maps) and 0 <= next_x < len(maps[0]) and maps[next_y][next_x] == 1:
                maps[next_y][next_x] = 0
                queue.append([next_y,next_x,cnt+1])
    
    return -1
    
def solution(maps):
    answer = bfs([0,0,1],maps)
    return answer

File: Programmers/JAVA/Programming_Maester/게임맵최단거리/게임맵최단거리.java
// 시간 효율성 테스트  통과 못하는 중 ... 


import java.util.LinkedList;
import java.util.Queue;
class Solution {
    class Location {
        int row;

        int col;  
        int distance; 
        public Location(int row,int col,int distance) {
            this.row =row;
            this.col = col;
     

    }
                     
    private int bfs(int[][] maps, Location start) {
        int[][] directs = {{1,0},{0,1},{-1,0},{0,-1}};
        int n = maps.length;
        int m = maps[0]

        q.off er(start);
    
        while(!q.is pty()) {   
            Location tmp = q.poll();
            int  r =  tmp.row, c=tmp.col, d =tmp.distance;
            maps[r][c] = 0 ; 
            for(int[] dir  :  directs) {
                in t dr = r + dir[0];  
                int dc = c+d i r[1];
                if(dr == n-1 && dc == m-1) {
                     return d+1;    
                } 
                if(0 <= dr && dr < n && 0<=dc &&  dc <   m){
                    if(maps[dr][dc] == 1) {
                        q.offer(new Location(dr,dc,d+1));
                    }
                }
            }
     

        return -1;
    }

        int answer = 0;  
        
        Location start = new Location(0,0,1);
        answer = bfs(maps, start);
        return answer;
    }
}